### PR TITLE
Add windows support

### PR DESCRIPTION
**Pull request by marcin-sielski**
This PR adds Windows support to Janus WebRTC Gateway. It supports
Linux and OS X as well. This version is based on v0.1.1 release.
This PR was created to discuss the changes introduced to the package.
It will be updated to the latest master if no major issues are found.

### CODE REVIEW

**Comment by lminiero** [comment]
Thanks for working on this! A quick glance showed an overwhelming number of changes... I'll try to review when I get back to the office in a week or so.

**Comment by marcin-sielski** [comment]
Thank, you...

**Comment by saghul** [comment]
@marcin-sielski what are you targeting, MinGW-w64?

File: README.md [hunk 1]
```diff
@@ -173,6 +172,121 @@ On Ubuntu/Debian:
...
+
+	mkdir mingw-w64-libmicrohttpd
+	cd mingw-w64-libmicrohttpd
+	wget https://raw.githubusercontent.com/Alexpux/MINGW-packages/master/mingw-w64-libmicrohttpd/PKGBUILD
+	makepkg-mingw -sLfi --noconfirm
+
+Sofia-SIP (x86, x86_64):
+
+	mkdir mingw-w64-sofia-sip-git
+	cd mingw-w64-sofia-sip-git
```
Comment (saghul) [inline README.md]: maybe just use git clone to simplify the steps?

File: janus.c [hunk 2]
```diff
@@ -2972,13 +2987,40 @@ static void janus_detect_local_ip(gchar *buf, size_t buflen) {
...
+	janus_log_timestamps = FALSE;
+	janus_log_colors = FALSE;
+	lock_debug = 0;
+	tasks = NULL;
+	sessions = NULL;
+	old_sessions = NULL;
+	sessions_watchdog_context = NULL;
+#ifdef _WIN32
+	gchar buf[MAX_PATH];
+	GetModuleFileName(NULL, buf, MAX_PATH);
```
Comment (saghul) [inline janus.c]: GetModuleFileNameW ?

File: os.c [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+
+#endif
+
+#ifndef HAVE_FREEIFADDRS
+
+void freeifaddrs(struct ifaddrs *ifp) {
+	struct ifaddrs *p, *q;
+
+	for(p = ifp; p; ) {
+	if (p->ifa_name)
```
Comment (saghul) [inline os.c]: style: alignment

File: README.md (hunk 1 above)
Comment (marcin-sielski) [inline README.md]: Hi,
Well, Yes, Actually this is the shortest and the most comfortable way to have it compiled. Once binary packages are avaialble in pacman repository it would need simply run "pacman -S --noconfirm mingw-w64-i686-sofia-sip-git"
Hope that helps.
Best Regards

**Comment by saghul** [comment]
I took the liberty of making a quick review :-) Awesome work @marcin-sielski!

File: README.md (hunk 1 above)
Comment (saghul) [inline README.md]: Sure, but in the meantime adding git to the required packages, cloning the repo and running makepkg in there should do, right?

Comment (marcin-sielski) [inline README.md]: makepkg-pkg together with PKGBUILD does all the magic (it clones the repos, applies patches, compiles and installs binaries). Hope that helps.

File: os.c (hunk 3 above)
Comment (marcin-sielski) [inline os.c]: True. I will fix this...

File: janus.c (hunk 2 above)
Comment (marcin-sielski) [inline janus.c]: Good point...

**Comment by marcin-sielski** [comment]
Thank you for the review

File: janus.c (hunk 2 above)
Comment (marcin-sielski) [inline janus.c]: Fixed

File: os.c (hunk 3 above)
Comment (marcin-sielski) [inline os.c]: Fixed

File: Makefile.am [hunk 4]
```diff
@@ -53,15 +59,21 @@ stream_DATA = $(NULL)
 recordingsdir = $(datadir)/janus/recordings
 recordings_DATA = $(NULL)
 
+if WINDOWS_OS
+INSTDIR = ../..
```
Comment (lminiero) [inline Makefile.am]: Why this? Shouldn't this be decided by the prefix you pass when configuring? Or does this work differently with MingW on Windows?

File: Makefile.am [hunk 5]
```diff
@@ -338,6 +407,7 @@ janus_pp_rec_CFLAGS = \
 
 janus_pp_rec_LDADD = \
 	$(POST_PROCESSING_LIBS) \
+	libjanus.la \
```
Comment (lminiero) [inline Makefile.am]: Why? The post-processor does not have the same dependencies as the Janus executable, most of the stuff there is not needed.

File: README.md [hunk 6]
```diff
@@ -173,6 +172,123 @@ On Ubuntu/Debian:
...
+
+	sc delete "Janus WebRTC Gateway"
+
+To start Janus WebRTC Windows Service execute:
+
+	sc start "Janus WebRTC Gateway"
+
+To stop Janus WebRTC Windows Service execute: 
+
+	sc stop "Janus WebRTC Gateway"
```
Comment (lminiero) [inline README.md]: Incredibly detailed documentation! Maybe too much... :smile: 
I'm wondering if the installation stuff should belong to a different .md file instead, so that the README is not too long. Anyway, just thinking out loud, and it is something we can worry about later.

File: Makefile.am (hunk 4 above)
Comment (marcin-sielski) [inline Makefile.am]: Hi Lorenzo, I hope you had great holidays. Thank you for your comment. This actually entire feature that enables to use relative paths within Janus code. It enables to create a package that is path independent so that sources can be compiled it in e.g. c:\msys64\mingw64 and binaries can be then installed in selected path by the user e.g. C:\Program Files\Janus WebRTC Gateway

File: apierror.h [hunk 7]
```diff
@@ -68,6 +70,6 @@
 /*! \brief Helper method to get a string representation of an API error code
  * @param[in] error The API error code
  * @returns A string representation of the error code */
-const char *janus_get_api_error(int error);
+shared const char *janus_get_api_error(int error);
```
Comment (lminiero) [inline apierror.h]: I've noticed you've added this `shared` pretty much everywhere. What is it for? Is it just an explicit way of saying a method is not `static`? Why is this needed on Windows? Is it safe to have for OSX and Linux? Just asking as I'm personally not familiar with it.

File: configure.ac [hunk 8]
```diff
@@ -20,10 +20,20 @@ darwin*)
 	CFLAGS="$CFLAGS -I/usr/local/opt/openssl/include"
 	LDFLAGS="$LDFLAGS -L/usr/local/lib -L/usr/local/opt/openssl/lib -L/opt/local/lib"
 	AM_CONDITIONAL([DARWIN_OS], true)
+  AM_CONDITIONAL([WINDOWS_OS], false)
```
Comment (lminiero) [inline configure.ac]: Note: indentation.

File: configure.ac [hunk 9]
```diff
@@ -20,10 +20,20 @@ darwin*)
...
+  CFLAGS="$CFLAGS -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 -D_POSIX_C_SOURCE -pipe -mwindows -mconsole"
+  LDFLAGS="$LDFLAGS -Wl,--enable-auto-image-base"
+  LIBS="-lws2_32 -liphlpapi"
+	AM_CONDITIONAL([DARWIN_OS], false)
+  AM_CONDITIONAL([WINDOWS_OS], true)
+	AM_CONDITIONAL([HAS_DTLS_WINDOW_SIZE], true)
+;;
 *)
 	AM_CONDITIONAL([DARWIN_OS], false)
+  AM_CONDITIONAL([WINDOWS_OS], false)
```
Comment (lminiero) [inline configure.ac]: Note: indentation.

File: configure.ac [hunk 10]
```diff
@@ -20,10 +20,20 @@ darwin*)
...
 	AM_CONDITIONAL([HAS_DTLS_WINDOW_SIZE], false)
 ;;
+mingw*)
+  CFLAGS="$CFLAGS -DWINVER=0x0A00 -D_WIN32_WINNT=0x0A00 -D_POSIX_C_SOURCE -pipe -mwindows -mconsole"
+  LDFLAGS="$LDFLAGS -Wl,--enable-auto-image-base"
+  LIBS="-lws2_32 -liphlpapi"
+	AM_CONDITIONAL([DARWIN_OS], false)
+  AM_CONDITIONAL([WINDOWS_OS], true)
+	AM_CONDITIONAL([HAS_DTLS_WINDOW_SIZE], true)
+;;
```
Comment (lminiero) [inline configure.ac]: Note: indentation.

File: apierror.h (hunk 7 above)
Comment (saghul) [inline apierror.h]: @lminiero it's defined here: https://github.com/meetecho/janus-gateway/pull/597/files#diff-b8c9b43cc2d006d6edf7b73b4a54fe10R22 but I have to admit it also threw me off the first time. Usuallt this is called something like JANUS_API or JANUS_EXTERN. IMHO something like that reads better.

Comment (lminiero) [inline apierror.h]: Ah thanks the clarification, I hadn't arrived at that point in my review yet, still looking at the first part. I agree that something like you suggest would avoid ambiguities, and potential issues with dependencies that define the same thing themselves.

Comment (marcin-sielski) [inline apierror.h]: Hi Lorenzo, "shared" is just #define defined in os.h
#ifdef _WIN32
#ifdef SHARED
#define shared __declspec(dllexport) 
#else
#define shared __declspec(dllimport)
#endif
#else
#define shared
#endif
It is empty on other OSes (no affect). It is used to mark particular object whether it should be imported to dll or exported.

File: configure.ac (hunk 9 above)
Comment (marcin-sielski) [inline configure.ac]: I will fix that

File: configure.ac (hunk 10 above)
Comment (marcin-sielski) [inline configure.ac]: I will fix that

File: configure.ac (hunk 8 above)
Comment (marcin-sielski) [inline configure.ac]: I will fix that

File: ice.c [hunk 11]
```diff
@@ -122,7 +125,12 @@ void janus_ice_force_rtcpmux(gboolean forced) {
 			JANUS_LOG(LOG_WARN, "Error creating RTCP component blackhole socket, using port %d instead\n", janus_force_rtcpmux_blackhole_port);
 			return;
 		}
+#ifdef _WIN32
+		u_long argp = 1;
+		ioctlsocket(blackhole, FIONBIO, &argp);
+#else
```
Comment (lminiero) [inline ice.c]: This is one of the things that frighten me, I'll most definitely break those anytime I add something sockety myself :wink:

File: apierror.h (hunk 7 above)
Comment (lminiero) [inline apierror.h]: Yep but as Saul pointed out, I'd use a different name for that. `shared` is too generic and risks to cause conflicts. Better something that explicitly belongs to the Janus namespace, e.g., `JANUS_EXTERN` or `JANUS_API` as Saul suggested, or something like that. Open to suggestions!

File: ice.c (hunk 11 above)
Comment (saghul) [inline ice.c]: You could probably create an internal function which sets the given fd in non-blocking mode, having it take care of the internals.

File: Makefile.am (hunk 5 above)
Comment (marcin-sielski) [inline Makefile.am]: I have noticed that there are many dependencies in transports/plugins/post-processor on the functions located in janus.c. I create shared object now based on that file.

File: README.md (hunk 6 above)
Comment (marcin-sielski) [inline README.md]: Ok, I will leave as it is for now.

File: janus.c [hunk 12]
```diff
@@ -36,38 +45,44 @@
...
 #else
 #define SHLIB_EXT ".so"
 #endif
+#endif
 
 
-static janus_config *config = NULL;
-static char *config_file = NULL;
-static char *configs_folder = NULL;
+shared janus_config *config;
```
Comment (lminiero) [inline janus.c]: Now this confuses me... why these static values are defined as shared instead? In this case, this would not result in them being static anymore on Linux.

File: apierror.h (hunk 7 above)
Comment (marcin-sielski) [inline apierror.h]: JANUS_SHARED?

File: os.c [hunk 13]
```diff
@@ -0,0 +1,535 @@
...
+
+		return -1;
+	}
+
+	return 0;
+}
+
+#endif
+
+#endif
```
Comment (lminiero) [inline os.c]: Missing newline here?

File: postprocessing/pp-h264.c [hunk 14]
```diff
@@ -78,7 +76,11 @@ int janus_pp_h264_create(char *destination) {
 	vStream->time_base = (AVRational){1, 90000};
 	vStream->codec->width = max_width;
 	vStream->codec->height = max_height;
+#if LIBAVCODEC_VER_AT_LEAST(53, 0)
+	vStream->codec->pix_fmt = AV_PIX_FMT_YUV420P;
+#else
```
Comment (lminiero) [inline postprocessing/pp-h264.c]: This looks like a generic fix we can add right away, I'll add it to master myself.

File: ice.c (hunk 11 above)
Comment (marcin-sielski) [inline ice.c]: I will fix that

File: postprocessing/pp-webm.c [hunk 15]
```diff
@@ -99,7 +100,11 @@ int janus_pp_webm_create(char *destination, int vp8) {
 	vStream->codec->time_base = (AVRational){1, fps};
 	vStream->codec->width = max_width;
 	vStream->codec->height = max_height;
+#if LIBAVCODEC_VER_AT_LEAST(53, 0)
+	vStream->codec->pix_fmt = AV_PIX_FMT_YUV420P;
+#else
```
Comment (lminiero) [inline postprocessing/pp-webm.c]: Same as above.

File: os.c (hunk 13 above)
Comment (marcin-sielski) [inline os.c]: I will fix that

**Comment by lminiero** [comment]
First quick review done, tremendous effort, thanks for this!

File: postprocessing/pp-h264.c (hunk 14 above)
Comment (marcin-sielski) [inline postprocessing/pp-h264.c]: There are many generic fixes there. This one of the exmaple. But without it I coud not compile it on Windows

Comment (lminiero) [inline postprocessing/pp-h264.c]: Yep, that's why I was thinking that some of the smaller generic fixes we might merge right away, either as a smaller PR or by applying them directly to the code, without waiting for this larger PR to be merged. This particular PR will need some more time, especially considering that as anticipated we have a couple of other PRs still waiting to be merged that may need a bit of refactoring in your effort too.

File: Makefile.am (hunk 4 above)
Comment (ploxiln) [inline Makefile.am]: That doesn't entirely make sense ...

I regularly compile janus in `/home/scratch/janus-gateway` and install to `/opt/janus`. You mean, the user can change the install dir after compile time? That's already sort of possible, by specifying the config file location on the janus command line (and then the config file specifies all the other paths). But this doesn't remove the need to do that, you only use INSTDIR for the location to install the sample config files ...

Presumably, what you would want to do for a "windows installer", is specify minimal `--prefix`, `--confdir` etc, and use the existing `DESTDIR` feature to install to a temporary "package root" folder, then the installer can put that anywhere, and the last necessary step is to customize the config file and launcher shortcut, from the installer, if that's what you really want to do. (You might also want to enable janus to use a relative path to the main config file, and relative paths within it, but this doesn't do that either.)

Comment (marcin-sielski) [inline Makefile.am]: Well, it does entirely make sense if you want to use 3rd party installer on Windows which is more common approach in that OS. I agree that paths could be changed after compilation but this make very limited use case on Windows. I also enabled relative path to the main config. Please refer to (https://github.com/marcin-sielski/janus-gateway/releases) binaries. Windows environment is by nature is slightly diffrent then Linux. Hope that helps.

**Comment by marcin-sielski** [comment]
Thanks Lorenzo!!!

File: apierror.h (hunk 7 above)
Comment (marcin-sielski) [inline apierror.h]: Changed to JANUS_API

File: configure.ac (hunk 8 above)
Comment (marcin-sielski) [inline configure.ac]: Fixed

File: configure.ac (hunk 9 above)
Comment (marcin-sielski) [inline configure.ac]: Fixed

File: configure.ac (hunk 10 above)
Comment (marcin-sielski) [inline configure.ac]: Fixed

File: ice.c (hunk 11 above)
Comment (marcin-sielski) [inline ice.c]: Fixed

File: os.c (hunk 13 above)
Comment (marcin-sielski) [inline os.c]: Fixed

**Comment by omichel** [comment]
I am very interested by this work as I am trying to setup an open source cross-platform WebRTC server (Linux, Windows and Mac OS X) that broadcasts video from an application producing images (OpenGL rendering). So far, I investigated OpenWebRTC and Janus (both combined with gstreamer) to achieve this goal, but none of them seemed to be well supported on Windows. Thanks to this PR, it seems Janus should now work on Windows... :smile: @marcin-sielski: I saw you also tried to compile OpenWebRTC on Windows, so I was wondering why you did this effort for Janus. Is Janus a better horse for setting-up such a cross-platform server?

**Comment by marcin-sielski** [comment]
Janus Gateway is definetly server side software. You could write Janus plugin (e.g. based on Streaming Plugin) using GStreamer Mutimedia framework to create full multimedia Server that takes as an input enything supported by GStreamer and have WebRTC output. OpenWebRTC works better for client side. Hope that helps. Anyway you may try Windows version of Janus published at https://github.com/marcin-sielski/janus-gateway/releases.

**Comment by radioman** [comment]
excellent effort! I wonder if visual studio would work?

**Comment by marcin-sielski** [comment]
Hi,

This is the first round of bringing WebRTC server software to Windows world.
The problems are always dependencies. Some of them can be already compiled using Visual
Studio compiler. I believe that this effort would significantly simplify second round - compile Janus with Visual Studio compiler.

I hope that helps

Best Regards

Marcin Sielski

File: janus.c (hunk 12 above)
Comment (marcin-sielski) [inline janus.c]: I have fixed that

**Comment by blackpan2** [comment]
@marcin-sielski I am using  the tagged release v0.2.0 (b543884), however when doing `janus.exe -v` I get janus 0.1.1 as the response. Is this is a bug or are we still working from v0.1.1 in the windows version? .

Thank you for bringing this to windows.

**Comment by marcin-sielski** [comment]
@blackpan2 Thank you for catching that. I have rebuilt Gateway from scratch. Should be fine now. Issue was related to the built itself and not to source code.

**Comment by blackpan2** [comment]
I just wanted to confirm, especially since it was saving v0.2.0 on startup. Thanks again for fixing it and integrating.

**Comment by abraiante** [comment]
It looks like this pull request is broken now. @marcin-sielski, any chance this will be worked on again so we can build janus in windows using master?

**Comment by marcin-sielski** [comment]
@abraiante Sure, I would it but I would like to have confirmation first from @lminiero when these changes could to be merged to the main line.

**Comment by lminiero** [comment]
As I said at the time, the main concern I have is maintaining all this. I don't have ways to compile the Windows code too (and I personally don't intend to set up a Windows environment), and I can't possibly ensure I don't break anything when I work on new features. Worrying about that could take me time I don't have and I'd rather dedicate to coding and fixing things.

If you can guarantee the changes will be manageable, no issue from me. You may want to wait a bit before taking on this effort, though, as in the next days I'll merge #784 which changes quite a bit in the SDP code and some plugins, and #799 will probably follow shortly after. Besides, it might be a good idea to base it on #403 instead of master, as that will be the next major merge after that (ideally by April, or May at the latest), and so you'd definitely be some steps ahead.

**Comment by lminiero** [comment]
Closing as a long time passed and way too many things changed in the meanwhile. An updated contribution as a new PR would be of course welcome. Thanks for this huge effort!

**Comment by caiiiycuk** [comment]
@marcin-sielski hi, thank you for great job. I trying to build latest version of Janus on windows. I successfully compiled it, and it starts without any visible error. However streaming plugin does not work properly. Did you tested streaming plugin with your changes, is it works for you?

My windows branch: https://github.com/caiiiycuk/janus-gateway/tree/windows

Can you give me any advice? More info:
https://groups.google.com/g/meetecho-janus/c/xPImgPDnxDI

**Comment by wangdemon** [comment]
> This PR adds Windows support to Janus WebRTC Gateway. It supports Linux and OS X as well. This version is based on v0.1.1 release. This PR was created to discuss the changes introduced to the package. It will be updated to the latest master if no major issues are found.


How to start the program?
